\protect\hypertarget{anchor}{}{}\emph{Lolium rigidum} (Annual ryegrass)
genome assembly, functional annotation, and landscape genomics across SE
Australia

Jefferson Paril, Gunjan Pandey, Emma B. Barnett, Rahul V. Rane,
Alexandre Fournier-Level

\textbf{Target:} Data Report in Frontiers in Genetics

\emph{This is the template:
}\href{https://www.frontiersin.org/articles/10.3389/fgene.2019.00889/full}{\emph{https://www.frontiersin.org/articles/10.3389/fgene.2019.00889/full}}

\protect\hypertarget{anchor-1}{}{}Introduction

\begin{itemize}
\tightlist
\item
  Lolium rigidum is remarkable in multiple ways.
\item
  It is the world's most herbicide resistant weed species. The first
  weed species reported to evolve resistance to glyphosate
\item
  Top weed species in wheat cropping - validate\ldots{}
\item
  One of the costliest weed species in Australia
\item
\end{itemize}

\protect\hypertarget{anchor-2}{}{}Biological description

Lolium rigidum, commonly known as annual ryegrass or rigid ryegrass or
Wimmera grass is an agriculturally-relevant weed species to study the
evolutionary quantitative genetics of herbicide resistance. L. rigidum
is considered as the world's most herbicide-resistant weed. It has
developed resistance to the highest number of herbicides with different
modes of action (i.e. 12 modes of action as of 2021) (Heap 2021). It is
the first weed species reported to have evolved resistance to glyphosate
(Powles et al. 1998). The combination of its fecundity and
cross-pollinated nature resulting in large genetically diverse
populations gives this weed species a high adaptive potential.

\emph{Lolium rigidum} (Gaudin, 1811) belongs to the grass family
Poaceae. Growth habits range from prostrate to erect, with erect being
the dominant type. It can grow up to 1 metre in height. Leaf blades are
5-25 cm long, 3-5 mm wide, green and glabrous (CABI 2017). Auricles are
small and narrow, while ligules are short and white to translucent. Leaf
sheaths are glabrous which can have purple colouration at the base.
Roots are extensive and fibrous. Inflorescence is a spike (length ≤ 30
cm) with 10-12 florets (10-25 mm long) in each spikelet (CABI 2017).
Each floret is generally awnless with equally-sized palea and lemma, and
three yellow anthers (Kloot 1983). It is a diploid with chromosome
number 2n=2x=14 (Terrell 1966; Monaghan 1980). It has an estimated
genome size of 2Gb similar to that of the closely-related forage crop
\emph{Lolium perenne} (Byrne et al. 2015). It is cross-compatible with
other members of the Lolium genus, i.e. \emph{L. multiflorum} and
\emph{L. perenne} (Kloot 1983). This genus is a complex of
cross-compatible species which can produce fertile hybrids and the
distinction between species is often blurred (Naylor 1960; Terrell 1966;
Kloot 1983).

It is a highly-competitive, self-incompatible, wind-pollinated, annual,
C3 weed species (Monaghan 1980; McCraw et al. 1983; CABI 2017). It can
produce up to 45,000 seeds m-2 in infested wheat fields (Gill 1996).
These seeds can have varying levels of dormancy ensuring the sustained
germination in the field and the replenishment of the soil seedbank
(Goggin et al. 2012). A density of 300 \emph{L. rigidum} plants m-2 can
cause significant reduction in rapeseed and cereal crop yields from
below 10\% to more than 50\% (Lemerle et al. 1995). Additionally, its
seeds can be infected by \emph{Clavibacter toxicus} which causes
livestock poisoning (Riley and McKay 1991; Ophel et al. 1993).

\protect\hypertarget{anchor-3}{}{}Centres of origin and diversity

It is native to the Mediterranean region, i.e. Europe and northern
Africa. It has spread across the temperate crop-growing regions around
the world. In the 19th century, it was introduced to Australia as a
forage crop (Kloot 1983). After years of artificial and natural
selection, it has adapted to local conditions and has become the major
weed in the wheat-growing regions of Australia (Reeves 1976; Medd et al.
1985; Powles and Matthews 1992).

\begin{itemize}
\tightlist
\item
  Potential other locations which they may take root in
\item
\item
  Genomic information can be leveraged to improve herbicide resistance
  and weed management in general.
\item
  The genome of a closely-related forage crop species, \emph{Lolium
  perenne}, is available but a better reference genome
  (chromosomal-level) and genome annotation (gene-level) are still
  lacking for this important herbicide resistant species.
\item
\item
  Here we assembled the genome of a glyphosate-resistant \emph{Lolium
  rigidum} plant collected from SE Australia.
\item
  We were able to generate a chromosomal-level assembly
\item
  We performed functional annotations and determined whether or not
  non-synonymous mutations are more common in this glyphosate resistant
  genome compared with that of \emph{Lolium perenne} and other various
  grass genomes.
\item
  Finally, we assess the pattern of genetic diversity of \emph{Lolium}
  populations across SE Australia
\item
\end{itemize}

\protect\hypertarget{anchor-4}{}{}Materials and Methods

\protect\hypertarget{anchor-5}{}{}Plant Sampling and Nucleic Acid
extraction

Sixty populations of weedy annual ryegrass were sampled across
Southeastern Australia (\protect\hyperlink{xpuavhcd9n22}{\emph{Figure
}}\protect\hyperlink{xpuavhcd9n22}{\emph{S1}}). For short-read,
whole-genome sequencing, total DNA was extracted from 100 mg of leaf
tissue using Plant DNeasy kits (Quigen, Germany). 60 independent DNA
extractions were performed, DNA was pooled and a sequencing library was
synthesised using NEB Ultra II DNA kit (New England Biolab, USA).

\begin{itemize}
\tightlist
\item
  ddRADseq with Illumina HiseqX
\end{itemize}

A single, highly glyphosate resistance plant originating from Wagga
Wagga (NSW, Australia, -35°0'16.01", 147°27'51.36") was selected to
build the \emph{Lolium rigidum} genome assembly. This individual
genotype was tissue-cultured to induce embryogenic calli for clonal
multiplication and maintenance. DNA extraction followed the same
protocol as above.

\begin{itemize}
\tightlist
\item
  Reference transcriptome point out from same genotype
\item
  Short-read sequencing no ddRAD just size-selection 200-700bp -
  Illumina HiseqX
\item
  Long-read sequencing with Promethion
\end{itemize}

\includegraphics[width=6.5728in,height=4.1398in]{Pictures/10000000000004D00000034A359391365B7EABBD.png}\protect\hypertarget{anchor-6}{}{}

\textbf{Supplementary Figure S1}. Sampling locations of 60 annual
ryegrass populations across Southeastern Australia collected in November
2018. Red pin refers to the location of the source population of the
individual from which the genome assembly was derived.

\protect\hypertarget{anchor-7}{}{}Whole Genome Sequencing

To assemble the genome, short- and long-read data were generated and
subsequently scaffolded using HiC scaffolding. DNA short-read sequencing
libraries were constructed using NEBNext Ultra II DNA Library Prep kit
for Illumina (NEB, USA), and sequenced using HiSeq X platform (Illumina,
Inc., San Diego, USA) with 150-bp paired-end reads. The reads were
filtered to remove 0.2\% adapter sequences using TrimGalore (v 0.6.6).
Long read sequencing was carried out on MinION and PromethION platforms.
The long-reads were basecalled using \emph{guppy }(v5.1) using the
\emph{dna\_r9.4.1\_450bps\_sup.cfg} model and trimmed using
\emph{Porechop} (v0.2.4) and filtered using \emph{filtlong} to obtain
41x highest quality reads. The long read sequences were assembled using
Flye (v2.9) and the minimum overlap parameter set to 6,000. Duplicate
contigs were purged using purge\_dups (v1.2.5). The long reads were
error corrected using \emph{Canu} (v2.2) and used to polish the contigs
three times using \emph{Racon} (v1.4.22). This was followed by three
rounds of short-read-based polishing using \emph{Polca} (\emph{MaSURCA}
v4.0.7) to obtain the final contig assembly. This assembly was assessed
using \emph{BUSCO} (v5) using the viridiplantae and poales lineages.

A Hi-C library was generated using 20mg of leaf tissue in the Arima HiC
kit following the manufacturer's instructions and sequenced on Illumina
Novaseq 6000 to obtain 500 million reads. Using the resulting genomic
topological information, the final contig assembly was scaffolded using
\emph{ALLHiC} (v1) and polished using \emph{3d-dna}. The assembly has
been submitted to NCBI under accession number
(\href{https://www.ncbi.nlm.nih.gov/nuccore/2206643664\#sequence_JAKKIG000000000.1}{\emph{SAMN25144995,
JAKKIG000000000}}).

\protect\hypertarget{anchor-8}{}{}Transcriptome sequencing and assembly

Clones from the reference plant were established through tissue culture
and grown under greenhouse conditions. Two independent samples of each
whole seedlings, roots, stems, leaves, inflorescence, meristems were
snap-frozen and ground and total RNA was extracted using Isolate II RNA
plant kit (Bioline, UK). RNA-sequencing libraries were individually
synthesised using NEBNext Ultra II stranded RNA library synthesis kits
(NEB, USA), indexed using the NEBNext Multiplex Oligos for Illumina
barcode kit (NEB, USA). Libraries were normalised and pooled to be
sequenced on an Illumina HiSeq X ten platform to generate 256,957,021
2x150bp reads. Raw, demultiplexed reads were error-corrected using
Rcorrector, adapters and low quality base pairs were trimmed using
TrimGalore v0.6.0. Ribosomal RNA was removed by discarding reads mapping
to the sequences deposited in SILVA v138.1 database using Bowtie2. After
filtering, 197,274,906 reads were used for de novo transcriptome
assembly using the DRAP workflow using the rice protein sequences (Oryza
sativa all peptides release 51) as guide and using both Trinity and
Oases as assembler. The two assemblies were then compacted into a single
compacted meta-assembly, filtered reads were then re-mapped against the
meta-assembly and transcripts with FPKM\textgreater1 were included in
the transcriptome.

\protect\hypertarget{anchor-9}{}{}Functional annotation and repetitive
element identification

\begin{itemize}
\item
  \textbf{Braker} pipeline native linux installations
\item
  Running from 2022/03/09
\item
  RNAseq informed
\item
  Plant protein database informed
\item
  Combined with \textbf{TSEBRA}
\item
  GeMoMa: Gene Model Mapper to predict identity of homologous sequences
  based on
\item
  Transposable elements identification and masking with
  \textbf{RepeatMasker}
\item
\item
  Comparative genomics

  \begin{itemize}
  \item
    \begin{quote}
    Using \textbf{GeMoMa} (genome model mapper) and TAIR10
    \emph{Arabidopsis thaliana} genome annotation and IRGSP-1.0
    \emph{Oryza sativa} genome annotations we generated two sets of
    mapped genome annotations, i.e. \emph{A.thaliana}-based and rice
    based gene names for the following species
    \end{quote}

    \begin{itemize}
    \tightlist
    \item
      Grasses and all C3: \emph{Lolium rigidm} (ours), \emph{Lolium
      perenne}, rice, and, rye
    \item
      Grass and C4 outgroup: maize
    \item
      Non-grass outgroups: \emph{A. thaliana }and \emph{Marachantia
      polymorpha}.
    \end{itemize}
  \item
    \begin{quote}
    Cluster gene families with \textbf{Panther HMM models}
    \end{quote}
  \item
    \begin{quote}
    Align gene families across species with \textbf{MAFFT}
    \end{quote}
  \item
    \begin{quote}
    Estimate divergence between species times using \textbf{MCMCTREE}
    and \textbf{TimeTree.org} fossil record estimates. Median divergence
    time between:
    \end{quote}

    \begin{itemize}
    \tightlist
    \item
      \emph{Marchantia polymorpha} and \emph{Arabidopsis thaliana}: 532
      MYA (465 - 533 MYA)
    \item
      \emph{Oryza sativa} and \emph{Zea mays}: 50 MYA (42 - 52 MYA)
    \item
      \emph{Arabidopsis thaliana }and \emph{Oryza sativa}: 160 MYA (115
      - 308 MYA)
    \item
      \emph{Arabidopsis thaliana }and \emph{Lolium rigidum: }160 MYA
      (115 - 308 MYA)
    \item
      \emph{Oryza sativa }and \emph{Lolium rigidum: }50 MYA (42 - 52
      MYA)
    \item
      \emph{Zea mays }and \emph{Lolium rigidum: }50 MYA (42 - 52 MYA)
    \item
      \emph{Lolium rigidum }and \emph{Lolium perenne: }1.65 MYA (0.00 -
      13.78 MYA)
    \end{itemize}
  \item
    \begin{quote}
    Dendrogram (phylo tree) all genes
    \end{quote}
  \item
    \begin{quote}
    syntenic gene comparisons for 4-fold degeneracy, i.e. tolerance for
    any point mutation (change into any of the 4 bases is fine) at the
    3rd position of a codon - probably look at 3-fold and 2-fold
    degeneracies as well
    \end{quote}
  \item
    \begin{quote}
    synteny pattern (gene family basis - core eukaryotic genes) between
    \emph{Lolium rigidum} and rye probably because they are both n=7
    which should look nice and symmetric :-)
    \end{quote}
  \end{itemize}
\item
\end{itemize}

\protect\hypertarget{anchor-10}{}{}Herbicide resistance genes

\begin{itemize}
\item
  Did we capture the genes coding for the target enzymes of the
  herbicides as well as detoxification genes which are likely to be
  recruited for herbicide resistance? \textbf{YES!}
\item
  Do we observe more non-synonymous than synonymous mutations in our
  resistant genotype vs Lolium perenne or rice or maize genomes? Using
  \textbf{PAML::codeml}
\item
  TST and NTSR (detox) gene expression heatmap per tissue

  \begin{itemize}
  \item
    \begin{quote}
    Quantity, gene family expansion (genomewide vs localised proximity
    with LTRs)
    \end{quote}
  \item
    \begin{quote}
    Divergence, non-synonymous substitutions
    \end{quote}
  \end{itemize}
\item
  Phylogenetic tree using TSR and detox genes in Lolium rigidum, rice,
  maize, and Arabidopsis -- BEAST software
\item
\end{itemize}

\protect\hypertarget{anchor-11}{}{}Landscape genomics

\begin{itemize}
\tightlist
\item
  Cluster and diversity analysis
\item
  Regress genetic distance against geographical distance
\item
  Recent bottleneck?
\item
  Isolation by distance?
\item
  High migration rate? Pattern correlated with combines and tractor
  ``migration''?
\end{itemize}

\protect\hypertarget{anchor-12}{}{}

\protect\hypertarget{anchor-13}{}{}Results

\protect\hypertarget{anchor-14}{}{}Assembly statistics

A total of 1,349,289,924 150bp paired-end Illumina reads were generated,
while 92.3Gb long reads with N\textsubscript{50}=16.2kb were generated
from MinION and PromethION (38.74Gb or 5,345,664 reads with
N\textsubscript{50}=14,897, and 53.62Gb or 3,697,296 reads with
N\textsubscript{50}=17,090, respectively). The Hi-C library generated
66Gb of raw sequence data.

This assembly was then assessed using BUSCO (v5) using the viridiplantae
{[}C:99.8\% {[}S:29.4\%, D:70.4\%{]}, F:0.2\%, M:0.0\%, n:425{]} and
poales lineages {[}C:97.2\% {[}S:32.4\%, D:64.8\%{]}, F:0.9\%, M:1.9\%,
n:4896{]}.

The final assembly is 2.44Gb with N\textsubscript{50} = 361.79Mb. In
Figure 1, we are only showing the 7 pseudo-chromosomes. The assembly has
been submitted to NCBI under accession number
(\href{https://www.ncbi.nlm.nih.gov/biosample/SAMN25144995}{\emph{SAMN25144995}},
\href{https://www.ncbi.nlm.nih.gov/nuccore/2206643664\#sequence_JAKKIG000000000.1}{\emph{JAKKIG000000000}})

\textbf{Table 1}. \emph{Lolium rigidum} genome assembly
(\href{https://www.ncbi.nlm.nih.gov/nuccore/2206643664\#sequence_JAKKIG000000000.1}{\emph{NCBI
GenBank: JAKKIG000000000.1}}) statistics.

\begin{longtable}[]{@{}ll@{}}
\toprule
& \\
\midrule
\endhead
Estimated genome size & 2.20 Gbp \\
Total size (bp) & 2.44 Gbp \\
Total contigs & 1,764 \\
N\textsubscript{50} & 0.36 Gbp \\
L\textsubscript{50} & 4 \\
N\textsubscript{90} & 0.36 Gbp \\
L\textsubscript{90} & 7 \\
Largest scaffold & 0.41 Gbp \\
Number of chromosomes & 7 \\
Total chromosome length & 2.35 Gbp \\
GC content & 44.76\% \\
Total length of Retroelements & 0.82 Gb (33.61\%) \\
Total length of DNA transposons & 0.09 Gb (3.84\%) \\
Number of gene models & 57,529 \\
Mean gene length & 3,623 bp \\
\bottomrule
\end{longtable}

\protect\hypertarget{anchor-15}{}{}Genomic features

The assembled \emph{Lolium rigidum} genome is 72.44\% interspersed
repeats. Transposable elements and repetitive sequences account for
33.61\% and 34.99\% of the genome, respectively. Among the transposable
elements, long terminal repeat (LTR) sequences were predominant (30.91\%
of the genome), composed of Copia (24.51\%) and Gyspy (6.40\%) LTR
families.

\includegraphics[width=5.6874in,height=4.9866in]{Pictures/10000201000007D0000007D0CC1B7D093C5C98C3.png}

\textbf{Figure 1}. Features of the \emph{Lolium rigidum} genome.
\textbf{a} - chromosome lengths: each tick is ×100Mb; \textbf{b} - GC
content heatmap ranging from 42\% to 47\% mean GC content per 2.35Mb
window; \textbf{c} distribution of Copia long terminal repeat (LTR)
retrotransposon family; \textbf{d} - distribution of Gypsy LTR
retrotransposon family. Chords at the centre represent syntenic
relationships between the top 5 orthogroups with the most paralogs in
the genome, where the colours match the colours of the chromosome most
of the paralogs per orthogroup are located.

\protect\hypertarget{anchor-16}{}{}Relationships and comparison with
other genomes

Insert tree with x-axis in time scale informed by tree.org
something\ldots{} + expansion/contraction gene counts, venn diagram of
common orthogroups + 4vDt thingy\ldots{}

\protect\hypertarget{anchor-17}{}{}Herbicide resistance genes

Comparative genetics using herbicide TSR and NTSR genes\ldots{}

\protect\hypertarget{anchor-18}{}{}Landscape genomics

Regress genetic distance against geographical distance\ldots{}

\protect\hypertarget{anchor-19}{}{}

\protect\hypertarget{anchor-20}{}{}Discussion

LTR and their expansion / proliferation / amplification in genomes has
been linked to domestication (Huang et al, 2017; Mo et al, 2015; Qin et
al, 2013; Qin et al, 2014).

\protect\hypertarget{anchor-21}{}{}References
