Jefferson Paril, Gunjan Pandey, Emma B. Barnett, Rahul V. Rane,
Alexandre Fournier-Level

\textbf{Target:} Data Report in Frontiers in Genetics

\emph{This is the template:
\href{https://www.frontiersin.org/articles/10.3389/fgene.2019.00889/full}{\uline{https://www.frontiersin.org/articles/10.3389/fgene.2019.00889/full}}}

\hypertarget{introduction}{%
\section{Introduction}\label{introduction}}

\begin{itemize}
\item
  \begin{quote}
  Lolium rigidum is remarkable in multiple ways.
  \end{quote}
\item
  \begin{quote}
  It is the world's most herbicide resistant weed species. The first
  weed species reported to evolve resistance to glyphosate
  \end{quote}
\item
  \begin{quote}
  Top weed species in wheat cropping - validate\ldots{}
  \end{quote}
\item
  \begin{quote}
  One of the costliest weed species in Australia
  \end{quote}
\item
\end{itemize}

\hypertarget{biological-description}{%
\subsubsection{Biological description}\label{biological-description}}

Lolium rigidum, commonly known as annual ryegrass or rigid ryegrass or
Wimmera grass is an agriculturally-relevant weed species to study the
evolutionary quantitative genetics of herbicide resistance. L. rigidum
is considered as the world's most herbicide-resistant weed. It has
developed resistance to the highest number of herbicides with different
modes of action (i.e. 12 modes of action as of 2021) (Heap 2021). It is
the first weed species reported to have evolved resistance to glyphosate
(Powles et al. 1998). The combination of its fecundity and
cross-pollinated nature resulting in large genetically diverse
populations gives this weed species a high adaptive potential.

\emph{Lolium rigidum} (Gaudin, 1811) belongs to the grass family
Poaceae. Growth habits range from prostrate to erect, with erect being
the dominant type. It can grow up to 1 meter in height. Leaf blades are
5-25 cm long, 3-5 mm wide, green and glabrous (CABI 2017). Auricles are
small and narrow, while ligules are short and white to translucent. Leaf
sheaths are glabrous which can have purple colouration at the base.
Roots are extensive and fibrous. Inflorescence is a spike (length ≤ 30
cm) with 10-12 florets (10-25 mm long) in each spikelet (CABI 2017).
Each floret is generally awnless with equally-sized palea and lemma, and
three yellow anthers (Kloot 1983). It is a diploid with chromosome
number 2n=2x=14 (Terrell 1966; Monaghan 1980). It has an estimated
genome size of 2Gb similar to that of the closely-related forage crop
\emph{Lolium perenne} (Byrne et al. 2015). It is cross-compatible with
other members of the Lolium genus, i.e. \emph{L. multiflorum} and
\emph{L. perenne} (Kloot 1983). This genus is a complex of
cross-compatible species which can produce fertile hybrids and the
distinction between species is often blurred (Naylor 1960; Terrell 1966;
Kloot 1983).

It is a highly-competitive, self-incompatible, wind-pollinated, annual,
C3 weed species (Monaghan 1980; McCraw et al. 1983; CABI 2017). It can
produce up to 45,000 seeds m-2 in infested wheat fields (Gill 1996).
These seeds can have varying levels of dormancy ensuring the sustained
germination in the field and the replenishment of the soil seedbank
(Goggin et al. 2012). A density of 300 \emph{L. rigidum} plants m-2 can
cause significant reduction in rapeseed and cereal crop yields from
below 10\% to more than 50\% (Lemerle et al. 1995). Additionally, its
seeds can be infected by Clavibacter toxicus which causes livestock
poisoning (Riley and McKay 1991; Ophel et al. 1993).

\hypertarget{centres-of-origin-and-diversity}{%
\subsubsection{Centres of origin and
diversity}\label{centres-of-origin-and-diversity}}

It is native to the Mediterranean region, i.e. Europe and northern
Africa. It has spread across the temperate crop-growing regions around
the world. In the 19th century, it was introduced to Australia as a
forage crop (Kloot 1983). After years of artificial and natural
selection, it has adapted to local conditions and has become the major
weed in the wheat-growing regions of Australia (Reeves 1976; Medd et al.
1985; Powles and Matthews 1992).

\begin{itemize}
\item
  \begin{quote}
  Potential other locations which they may take root in
  \end{quote}
\item
\item
  \begin{quote}
  Genomic information can be leveraged to improve herbicide resistance
  and weed management in general.
  \end{quote}
\item
  \begin{quote}
  The genome of a closely-related forage crop species, \emph{Lolium
  perenne}, is available but a better reference genome
  (chromosomal-level) and genome annotation (gene-level) are still
  lacking for this important herbicide resistant species.
  \end{quote}
\item
\item
  \begin{quote}
  Here we assembled the genome of a glyphosate-resistant \emph{Lolium
  rigidum} plant collected from SE Australia.
  \end{quote}
\item
  \begin{quote}
  We were able to generate a chromosomal or near-chomosomal-level
  assembly?
  \end{quote}
\item
  \begin{quote}
  We performed functional annotations and determined whether or not
  non-synonymous mutations are more common in this glyphosate resistant
  genome compared with that of \emph{Lolium perenne} and other various
  grass genomes.
  \end{quote}
\item
  \begin{quote}
  Finally, we assess the pattern of genetic diversity of \emph{Lolium}
  populations across SE Australia
  \end{quote}
\item
\end{itemize}

\hypertarget{materials-and-methods}{%
\section{Materials and Methods}\label{materials-and-methods}}

\hypertarget{plant-sampling-and-nucleic-acid-extraction}{%
\subsection{Plant Sampling and Nucleic Acid
extraction}\label{plant-sampling-and-nucleic-acid-extraction}}

Sixty populations of weedy annual ryegrass were sampled across
Southeastern Australia (\protect\hyperlink{xpuavhcd9n22}{\uline{Figure}
\uline{S1}}). For short-read, whole-genome sequencing, total DNA was
extracted from 100 mg of leaf tissue using Plant DNeasy kits (Quigen,
Germany). 60 independent DNA extractions were performed, DNA was pooled
and a sequencing library was synthesized using NEB Ultra II DNA kit (New
England Biolab, USA).

\begin{itemize}
\item
  \begin{quote}
  ddRADseq with Illumina HiseqX
  \end{quote}
\end{itemize}

A single, highly glyphosate resistance plant originating from Wagga
Wagga (NSW, Australia, -35°0'16.01", 147°27'51.36") was selected to
build the \emph{Lolium rigidum} genome assembly. This individual
genotype was tissue-cultured to induce embryogenic calli for clonal
multiplication and maintenance. DNA extraction followed the same
protocol as above.

\begin{itemize}
\item
  \begin{quote}
  Reference transcriptome point out from same genotype
  \end{quote}
\item
  \begin{quote}
  Short-read sequencing no ddRAD just size-selection 200-700bp -
  Illumina HiseqX
  \end{quote}
\item
  \begin{quote}
  Long-read sequencing with Promethion
  \end{quote}
\end{itemize}

\includegraphics[width=6.57292in,height=4.13931in]{media/image1.png}

\textbf{Supplementary Figure S1}. Sampling locations of 60 annual
ryegrass populations across Southeastern Australia collected in November
2018. Red pin refers to the location of the source population of the
individual from which the genome assembly was derived.

\hypertarget{whole-genome-sequencing}{%
\subsection{Whole Genome Sequencing}\label{whole-genome-sequencing}}

Genome sequencing:

\begin{itemize}
\item
  \begin{quote}
  llumina Hiseq X ten:
  \end{quote}
\item
  \begin{quote}
  MinIOn sequencing + Promethion (\textgreater10kb reads kept)
  \end{quote}
\item
  \begin{quote}
  Guppy 5.1+ (model: dna\_r9.4.1\_450bps\_sup.cfg \textbar{} sed
  `s/sup/hac/g' faster)
  \end{quote}
\item
  \begin{quote}
  Hi-C
  \end{quote}
\item
  \begin{quote}
  MinION + Promethion merged, assembled, polished with Illumina,
  deduplication, BUSCO completeness, and chromosome-level scaffolding
  with Hi-C (1 - 5 November expect the stitched assembly)
  \end{quote}
\item
  \begin{quote}
  QC steps
  \end{quote}
\item
\end{itemize}

\hypertarget{transcriptome-sequencing-and-assembly}{%
\subsection{Transcriptome sequencing and
assembly}\label{transcriptome-sequencing-and-assembly}}

Clones from the reference plant were established through tissue culture
and grown under greenhouse conditions. Two independent samples of each
whole seedlings, roots, stems, leaves, inflorescence, meristems were
snap-frozen and ground and total RNA was extracted using Isolate II RNA
plant kit (Bioline, UK). RNA-sequencing libraries were individually
synthesised using Ultra II stranded RNA library synthesis kits (NEB,
USA), indexed using the XXX barcode kit (NEB, USA). Libraries were
normalised and pooled to be sequenced on an Illumina HiSeq X ten
platform to generate 256,957,021 2x150bp reads. Raw, demultiplexed reads
were error-corrected using Rcorrector, adapters and low quality base
pairs were trimmed using TrimGalore v0.6.0. Ribosomal RNA was removed by
discarding reads mapping to the sequences deposited in SILVA v138.1
database using Bowtie2. After filtering, 197,274,906 reads were used for
de novo transcriptome assembly using the DRAP workflow using the rice
protein sequences (Oryza sativa all peptides release 51) as guide and
using both Trinity and Oases as assembler. The two assemblies were then
compacted into a single compacted meta-assembly, filtered reads were
then re-mapped against the meta-assembly and transcripts with
FPKM\textgreater1 were included in the transcriptome.

\hypertarget{functional-annotation-and-repetitive-element-identification}{%
\subsection{Functional annotation and repetitive element
identification}\label{functional-annotation-and-repetitive-element-identification}}

\begin{itemize}
\item
  \begin{quote}
  Augustus functional annotation
  \end{quote}
\item
  \begin{quote}
  Breaker2 pipeline (containerise with Docker vs Conda environment?)
  \end{quote}
\item
  \begin{quote}
  Blast2Go?
  \end{quote}
\item
  \begin{quote}
  Let NCBI do the annotation automatically most probably???
  \end{quote}
\item
  \begin{quote}
  NCBI: take our genome, use all SRA protein proteins from the NCBI
  database and do their annotation. And will finish within a week or
  two?
  \end{quote}
\item
\item
  \begin{quote}
  \textbf{RepeatModeler} (RECON, RepeatScout, and LTR\_retriever) to
  generate the repeat sequences database with the transposable elements
  database Dfam version 3.5 (2021 version)
  \end{quote}
\item
  \begin{quote}
  \textbf{RepeatMasker} to pinpoint the location of the repeat sequences
  in the genome
  \end{quote}
\item
  \begin{quote}
  Both these tools are dependent upon multiple tools (see the
  \href{https://github.com/jeffersonfparil/Lolium_rigidum_genome_assembly_and_annotation/blob/main/repetitive_elements_identification.md}{\uline{code}}
  for details)
  \end{quote}
\item
\end{itemize}

\hypertarget{herbicide-resistance-genes}{%
\subsection{Herbicide resistance
genes}\label{herbicide-resistance-genes}}

\begin{itemize}
\item
  \begin{quote}
  Did we capture the genes coding for the target enzymes of the
  herbicides as well as detoxification genes which are likely to be
  recruited for herbicide resistance?
  \end{quote}
\item
  \begin{quote}
  Do we observe more non-synonymous than synonymous mutations in our
  resistant genotype vs Lolium perenne or rice or maize genomes?
  \end{quote}
\item
\end{itemize}

\hypertarget{landscape-genomics}{%
\subsection{Landscape genomics}\label{landscape-genomics}}

\begin{itemize}
\item
  \begin{quote}
  Cluster and diversity analysis
  \end{quote}
\item
  \begin{quote}
  Regress genetic distance against geographical distance
  \end{quote}
\item
  \begin{quote}
  Recent bottleneck?
  \end{quote}
\item
  \begin{quote}
  Isolation by distance?
  \end{quote}
\item
  \begin{quote}
  High migration rate? Pattern correlated with combines and tractor
  ``migration''?
  \end{quote}
\end{itemize}

\hypertarget{section}{%
\section{\texorpdfstring{\hfill\break
}{ }}\label{section}}

\hypertarget{results}{%
\section{Results}\label{results}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\item
  \begin{quote}
  Assembly characterisation: Circos figure showing the features of the
  genome
  \end{quote}

  \begin{itemize}
  \item
    \begin{quote}
    length of each pseudo chromosome
    \end{quote}
  \item
    \begin{quote}
    distribution of retrotransposons
    \end{quote}
  \item
    \begin{quote}
    heatmap of GC content
    \end{quote}
  \item
    \begin{quote}
    whole-genome duplication events showing the syntenic relationships
    between blocks containing multiple (e.g. 10) paralogous gene pairs
    \end{quote}
  \end{itemize}
\end{enumerate}

\includegraphics[width=6.5in,height=6.33934in]{media/image3.png}

\textbf{Figure 1}. Features of the \emph{Lolium rigidum} genome.
\textbf{a} - chromosome lengths: each tick is ×100Mb; \textbf{b} - GC
content heatmap ranging from 42\% to 47\% mean GC content per 2.35Mb
window; \textbf{c} distribution of Ty1-Copia long terminal repeat (LTR)
retrotransposon family; \textbf{d} - distribution of Ty1-Gypsy LTR
retrotransposon family.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{1}
\item
  \begin{quote}
  Relationships and comparison with other genomes
  \end{quote}

  \begin{itemize}
  \item
    \begin{quote}
    Compare with:
    \end{quote}

    \begin{itemize}
    \item
      \begin{quote}
      C3: \emph{Lolium perenne}, rice, rye, barley \& \emph{Aegilops
      tauschii}
      \end{quote}
    \item
      \begin{quote}
      C4: maize, sorghum, \& \emph{Cynodon transvaalensis}
      \end{quote}
    \item
      \begin{quote}
      Non-grass outgroup: Arabidopsis, Soybean \& Marchantia
      \end{quote}
    \item
      \begin{quote}
      Lolium perenne assembly is too fragmented and will be very close
      to L rigidum
      \end{quote}
    \end{itemize}
  \item
    \begin{quote}
    Dendrogram
    \end{quote}
  \item
    \begin{quote}
    syntenic gene comparisons for 4-fold degeneracy, i.e. tolerance for
    any point mutation (change into any of the 4 bases is fine) at the
    3rd position of a codon - probably look at 3-fold and 2-fold
    degeneracies as well
    \end{quote}
  \item
    \begin{quote}
    synteny pattern (gene family basis - core eukaryotic genes) between
    \emph{Lolium rigidum} and rye probably because they are both n=7
    which should look nice and symmetric :-)
    \end{quote}
  \end{itemize}
\item
  \begin{quote}
  Herbicide resistance genes
  \end{quote}

  \begin{itemize}
  \item
    \begin{quote}
    TST and NTSR (detox) gene expression heatmap per tissue
    \end{quote}
  \item
    \begin{quote}
    Phylogenetic tree using TSR and detox genes in Lolium rigidum, rice,
    maize, and Arabidopsis -- BEAST software
    \end{quote}
  \end{itemize}
\item
  \begin{quote}
  Landscape genomics
  \end{quote}
\end{enumerate}

\includegraphics[width=6.5in,height=8.125in]{media/image2.png}

\hypertarget{section-1}{%
\section{\texorpdfstring{\hfill\break
}{ }}\label{section-1}}

\hypertarget{discussion}{%
\section{Discussion}\label{discussion}}

\hypertarget{references}{%
\section{References}\label{references}}
